README.md
# SiriDB Admin Tool
![alt SiriDB Admin](/siridb-admin.png?raw=true)

Tool for managing SiriDB service accounts and databases. SiriDB-Admin can be used both by command-line arguments and a graphical web-interface.

---------------------------------------
  * [Requirements](#requirements)
  * [Installation](#installation)
    * [Pre-compiled](#pre-compiled)
    * [Compile from source](#compile-from-source)
  * [Graphical Web Interface](#graphical-web-interface)
  * [Command-line arguments](#command-line-arguments)
    * [Service accounts](#service-accounts)
      * [List service accounts](#list-service-accounts)
      * [New service account](#new-service-account)
      * [Change password](#change-password)
      * [Remove service account](#remove-service-account)
    * [Databases](#databases)
      * [List databases](#list-databases)
      * [New database](#new-database)
      * [New replica](#new-replica)
      * [New pool](#new-pool)
      * [Drop database](#drop-database)

---------------------------------------
## Requirements
SiriDB Admin only works with SiriDB version 2.0.16 and higher. For older versions of SiriDB the manage tool is required and can be found at https://github.com/SiriDB/siridb-manage

## Installation
SiriDB Admin Tool can be compiled from source or, for most systems, you can simply download a pre-compiled binary.

### Pre-compiled
Go to https://github.com/SiriDB/siridb-admin/releases/latest and download the binary for your system.
In this documentation we refer to the binary as `siridb-admin`. On Linux/OSX it might be required to set the execution flag:
```
$ chmod +x siridb-admin_X.Y.Z_OS_ARCH.bin
```

You might want to copy the binary to /usr/local/bin and create a symlink like this:
```
$ sudo cp siridb-admin_X.Y.Z_OS_ARCH.bin /usr/local/bin/
$ sudo ln -s /usr/local/bin/siridb-admin_X.Y.Z_OS_ARCH.bin /usr/local/bin/siridb-admin
```
Note: replace `X.Y.Z_OS_ARCH` with your binary, for example `1.2.0_linux_amd64`

### Compile from source
> Before compiling from source make sure **go**, **npm** and **git** are installed and your [$GOPATH](https://github.com/golang/go/wiki/GOPATH) is set.

Clone the project using git. (we assume git is installed)
```
git clone https://github.com/SiriDB/siridb-admin
```

Make sure less is installed:
```
$ sudo npm install -g less less-plugin-clean-css
```

The gobuild.py script can be used to build the binary:
```
$ ./gobuild.py -i -l -w -b -p
```

Or, if you want the development version which uses original files from /build and /static instead of build-in files:
```
$ ./gobuild.py -i -l -w -b -d
```

## Graphical Web Interface
A web interface can be started with the following command:
```
siridb-admin --http
```
This will start a webserver listening on port 8080. You can now access the graphical interface by opening url http://localhost:8080 in your favorite web-browser.

Note: you might want to use a different port for the webserver. This can be done using the `-O` or `--port` argument flag. For example:
```
siridb-admin --http --port 5050
```

## Command-line arguments
SiriDB Admin can be used by just using command line arguments. This can be useful in case you want to create and extend a SiriDB database using a script.

All commands start with the following arguments:
```
siridb-admin -u <service_account> [-p <password>] [-s <siridb_server>] <command> ...
```
The service account is a required argument. By default the `sa` user with password `siri` is installed on a SiriDB server and this account will be used in most examples in this documentation.

If the `-p` flag with the service account password is not given, the tool will ask for the service account password.

The other optional argument flag `-s` can be used to provide a SiriDB server address and optional port. If not given then the default `localhost:9000` is used. If we, for example, have installed SiriDB on a server with hostname `siridb01.foo.local`, we can simply use argument `-s siridb01.foo.local`. Note that adding `:9000` is not required since port 9000 is the default port.

It is also possible to use an IPv4 or IPv6 address instead of a hostname. In case you want to use both an IPv6 address and use an alternative port then do not forget to put the IPv6 address between braces. For example: `[::1]:5050` is a valid IPv6 address and port.

If something goes wrong you should usually get an error message with some information. More information can be printed when the `--verbose` flag is used. The `--verbose` flag works together with all arguments including the `--http` argument.

### Service accounts
Service accounts are used for managing databases. A service account is a user on a SiriDB server that never has access to a SiriDB database. We call users with access to a database 'database users'. We keep these accounts seperate because database accounts exist in a database which possible extends over multiple SiriDB servers.

### List service accounts
Get all service accounts on siridb01.foo.local using service account sa with password siri:
```
siridb-admin -u sa -p siri -s siridb01.foo.local get-accounts
```

#### New service account
A new service account can be created using the command below.
```
siridb-admin -u sa -s siridb01.foo.local new-account bob passwd4bob
```
This command will ask for the `sa` password and then create a service account `bob` with password `passwd4bob` on a SiriDB server with hostname `siridb01.foo.local`.

#### Change password
A password can be changed using the following command:
```
siridb-admin -u sa -p siri -s siridb01.foo.local:9000 change-password newpassw0rd
```
This will change the password for the `sa` service account on SiriDB server `siridb01.foo.local` from `siri` to `newpassw0rd`.

#### Remove service account
Service accounts can be removed using the following command:
```
siridb-admin -u sa -p siri -s siridb01.foo.local drop-account bob
```
This command will remove service account `bob` from SiriDB server `siridb01.foo.local`.

### Databases
SiriDB Admin can be used to create a new database or extend an existing database with a new pool or replica. Note that databases can not be removed with this tool. As long as a database exists only on one SiriDB server you can remove the database by stopping the siridb-server process and then remove the database directory. All database directories can be found in the database path which can be configured in the siridb configuration file.

### List databases
Get all existing databases on a SiriDB server siridb01.foo.local using service account sa with password siri:
```
siridb-admin -u sa -p siri -s siridb01.foo.local get-databases
```

### New database
A new database can be created with the following syntax:
```
siridb-admin -u <service_account> [flags] new-database
  -d, --db-name=DB-NAME        Database name. (Must be at least 2 and at most 20 characters. First
                               character must be a letter. Last character must be a letter or number.
                               In between letters, numbers, hyphen and underscores are allowed)
  -t, --time-precision="ms"    Time precision for the new database. Supported time precissions are
                               s (second), ms (millisecond), us (microsecond) and ns (nanosecond).
                               This value cannot be changed once the database is created.
  -b, --buffer-size=1024       Buffer size for the new database. Each series uses a buffer space in
                               both memory and disk before points are actually written to shards.
                               This value cannot be changed once the database is created.
  -N, --duration-num="1w"      Number duration for the new database. Points are written to shards and
                               each shard has points for a specific time range. The size or time
                               window can be chosen but not changed once the database is created. For
                               example: the value '1w' will create shards holding points for 1 week.
  -L, --duration-log="1d"      Log duration for the new database. Like numeric duration but then for
                               log values. At the moment log values are not supported by SiriDB but
                               this will be implemented in a future release.

```
For example:
```
siridb-admin -u sa -p siri -s siridb01.foo.local new-database -d dbexample -t s
```
This will create database `dbexample` on SiriDB server `siridb01.foo.local` with a *second* time precision.

>Note: each new database will be created with a default database user `iris` and password `siri`.

### New replica
The following syntax can be used to create a new replica:
```
siridb-admin -u <service_account> [flags] new-replica
  -d, --db-name=DB-NAME          Database name where you want to add the new replica to.
  -U, --db-user=DB-USER          User with full privileges to the database.
  -P, --db-password=DB-PASSWORD  Password for the database user.
  -S, --db-server=DB-SERVER      SiriDB server address[:port]. Can be any server from the database
                                 you want to add a new replica to.
  -o, --pool=POOL                Pool number which you want to create the replica for.
  -f, --force                    Suppress warning message.
```
For example:
```
siridb-admin -u sa -p siri -s siridb02.foo.local new-replica -d dbexample -U iris -P siri -S siridb01.foo.local --pool 0
```
This will ask for confirmation and then create a replica on SiriDB server `siridb02.foo.local` for pool `0` in database `dbexample`.

### New pool
The following syntax can be used to create a new pool:
```
siridb-admin -u <service_account> [flags] new-pool
  -d, --db-name=DB-NAME          Database name where you want to add the new pool to.
  -U, --db-user=DB-USER          User with full privileges to the database.
  -P, --db-password=DB-PASSWORD  Password for the database user.
  -S, --db-server=DB-SERVER      SiriDB server address[:port]. Can be any server from the database
                                 you want to add a new pool to.
  -f, --force                    Suppress warning message.
```
For example:
```
siridb-admin -u sa -p siri -s siridb03.foo.local new-pool -d dbexample -U iris -P siri -S siridb01.foo.local
```
This will ask for confirmation and then create a new pool on SiriDB server `siridb03.foo.local` for database `dbexample`. Pool id's will be incremented automatically so if the database only had pool `0` then the new pool will have id `1`.

This will ask for confirmation and then create a replica on SiriDB server `siridb02.foo.local` for pool `0` in database `dbexample`.

### Drop database
>Note: requires at least SiriDB version 2.0.31

The following syntax can be used to drop a database:
```
siridb-admin -u <service_account> [flags] drop-database
  -d, --db-name=DB-NAME          Database name where you want to add the new pool to.
  -i, --ignore-offline           Usually the database will be removed from all servers in the
                                 cluster and therefore an error is returned in case one or more
                                 servers are offline. This flag can be used to ignore offline servers.
  -f, --force                    Suppress warning message.
```
For example:
```
siridb-admin -u sa -p siri -s siridb01.foo.local drop-database -d dbexample
```
This will ask for confirmation and then remove the database from all servers.



ChangeLog
2019.01.07, Version 1.2.0

  * Added drop database option. (issue #10)
  * Update SiriDB GO Connector
  * Use Vlow instead of Reflux

2017.11.14, Version 1.1.3

  * Update SiriDB GO Connector

2017.05.04, Version 1.1.2

  * Fixed accepting IPv6 SiriDB server address from command line. (issue #5)
  * Freeze npm package versions.

2017.04.06, Version 1.1.1

  * Fixed verbose logging. (issue #1)
  * Fixed honoring the force argument when creating a new replica. (issue #2)

2017.04.06, Version 1.1.0

  * Renamed some command line arguments.
  * Print error message if the web server cannot start listening.

2017.04.03, Version 1.0.0

  * Initial Version, providing both full command line support and a graphical
    interface for managing service accounts and databases.


handlersprod.go
// +build !debug

package main

import "net/http"

func handlerMain(w http.ResponseWriter, r *http.Request) {
	w.Write(FileIndexHTML)
}

func handlerJsBundle(w http.ResponseWriter, r *http.Request) {
	w.Write(FileBundleMinJS)
}

func handlerFaviconIco(w http.ResponseWriter, r *http.Request) {
	w.Write(FileFaviconICO)
}

func handlerBootstrapCSS(w http.ResponseWriter, r *http.Request) {
	w.Write(FileBootstrapMinCSS)
}

func handlerLayout(w http.ResponseWriter, r *http.Request) {
	w.Write(FileLayoutMinCSS)
}

func handlerFontAwesomeMinCSS(w http.ResponseWriter, r *http.Request) {
	w.Write(FileFontAwesomeMinCSS)
}

func handlerFontsFaOTF(w http.ResponseWriter, r *http.Request) {
	w.Write(FileFontAwesomeOTF)
}

func handlerFontsFaEOT(w http.ResponseWriter, r *http.Request) {
	w.Write(FileFontawesomeWebfontEOT)
}

func handlerFontsFaSVG(w http.ResponseWriter, r *http.Request) {
	w.Write(FileFontawesomeWebfontSVG)
}

func handlerFontsFaTTF(w http.ResponseWriter, r *http.Request) {
	w.Write(FileFontawesomeWebfontTTF)
}

func handlerFontsFaWOFF(w http.ResponseWriter, r *http.Request) {
	w.Write(FileFontawesomeWebfontWOFF)
}

func handlerFontsFaWOFF2(w http.ResponseWriter, r *http.Request) {
	w.Write(FileFontawesomeWebfontWOFF2)
}

func handlerSiriDBLargePNG(w http.ResponseWriter, r *http.Request) {
	w.Write(FileSiriDBLargePNG)
}

func handlerSiriDBSmallPNG(w http.ResponseWriter, r *http.Request) {
	w.Write(FileSiriDBSmallPNG)
}

func handlerLoaderGIF(w http.ResponseWriter, r *http.Request) {
	w.Write(FileLoaderGIF)
}


handlerstest.go
// +build debug

package main

import (
	"fmt"
	"io/ioutil"
	"net/http"
)

func handlerMain(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./src/index.html")
}

func handlerJsBundle(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./build/bundle.js")
}

func handlerFaviconIco(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/favicon.ico")
}

func handlerBootstrapCSS(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/css/bootstrap.css")
}

func handlerLayout(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./build/layout.css")
}

func handlerFontAwesomeMinCSS(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/css/font-awesome.min.css")
}

func handlerFontsFaOTF(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/fonts/FontAwesome.otf")
}

func handlerFontsFaEOT(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/fonts/fontawesome-webfont.eot")
}

func handlerFontsFaSVG(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/fonts/fontawesome-webfont.svg")
}

func handlerFontsFaTTF(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/fonts/fontawesome-webfont.ttf")
}

func handlerFontsFaWOFF(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/fonts/fontawesome-webfont.woff")
}

func handlerFontsFaWOFF2(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/fonts/fontawesome-webfont.woff2")
}

func handlerSiriDBLargePNG(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/img/siridb-large.png")
}

func handlerSiriDBSmallPNG(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/img/siridb-small.png")
}

func handlerLoaderGIF(w http.ResponseWriter, r *http.Request) {
	handleFileRequest(w, "./static/img/loader.gif")
}

func handleFileRequest(w http.ResponseWriter, fn string) {
	b, err := ioutil.ReadFile(fn)
	if err == nil {
		_, err = w.Write(b)
	} else {
		w.WriteHeader(http.StatusInternalServerError)
		_, err = fmt.Fprintf(w, "Internal server error: %s", err)
	}
	if err != nil {
		fmt.Println(err)
	}
}

func init() {
	fmt.Println("DEBUG MODE: using original template files...")
}


src/webpack.config.js
// Production:
//
//      NODE_ENV='production' ./node_modules/.bin/webpack -p
//
var webpack = require('webpack');
var path = require('path');

const BUILD_DIR = path.resolve(__dirname, '../build');
const APP_DIR = path.resolve(__dirname, '');

var config = {
    cache: true,
    entry: APP_DIR + '/Components/index.jsx',
    output: {
        path: BUILD_DIR,
        filename: process.env.NODE_ENV === 'production' ? 'bundle.min.js' : 'bundle.js',
    },
    module: {
        rules: [
            {
                test: /\.jsx?/,
                include: APP_DIR,
                loader: 'babel-loader'
            }
        ]
    },
    plugins: [
        new webpack.DefinePlugin({
            // A common mistake is not stringifying the "production" string.
            'process.env': {
                'NODE_ENV': JSON.stringify(process.env.NODE_ENV)
            }
        }),
    ]
};

if (process.env.NODE_ENV === 'production') {
    config.plugins.push(
        new webpack.optimize.UglifyJsPlugin({
            compress: {
                warnings: false,
                screw_ie8: false
            }
        })
    );
}

module.exports = config;

admin.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"strconv"

	siridb "github.com/SiriDB/go-siridb-connector"
	"github.com/astaxie/beego/session"
	"github.com/howeyc/gopass"

	"strings"

	kingpin "gopkg.in/alecthomas/kingpin.v2"
)

// AppVersion exposes version information
const AppVersion = "1.2.0"

var (
	xApp      = kingpin.New("siridb-admin", "Tool for creating and expanding SiriDB databases.")
	xAccount  = xApp.Flag("user", "Service account name for connecting to the SiriDB server. (ignored with the --http flag)").Short('u').Required().String()
	xPassword = xApp.Flag("password", "Password for your service account.").Short('p').String()
	xServer   = xApp.Flag("server", "Server address[:port] for the SiriDB server.").Short('s').Default("localhost:9000").String()
	xHTTP     = xApp.Flag("http", "Start a webserver for a graphical user interface. (only --port flag is parsed, other arguments/flags are ignored)").Bool()
	xHTTPPort = xApp.Flag("port", "Specific port for the http webserver.").Short('O').Default("8080").Int()
	xVerbose  = xApp.Flag("verbose", "Enable verbose logging.").Bool()
	xVersion  = xApp.Flag("version", "Print version information and exit.").Bool()

	xGetVersion   = xApp.Command("get-version", "Returns SiriDB server version information.")
	xGetAccounts  = xApp.Command("get-accounts", "Returns all service accounts.")
	xGetDatabases = xApp.Command("get-databases", "Returns all databases on the SiriDB server.")

	xNewAccount = xApp.Command("new-account", "Create a new service account.")
	xNaAccount  = xNewAccount.Arg("name", "Name for the new service account.").Required().String()
	xNaPassword = xNewAccount.Arg("password", "Password for the new service account.").Required().String()

	xDropAccount = xApp.Command("drop-account", "Remove a service account.")
	xDaAccount   = xDropAccount.Arg("name", "Account name which you want to drop.").Required().String()

	xDropDatabase    = xApp.Command("drop-database", "Remove a database.")
	xDdDatabase      = xDropDatabase.Flag("db-name", "Database name to drop.").Short('d').Required().String()
	xDdIgnoreOffline = xDropDatabase.Flag("ignore-offline", "Continue even if servers are offline.").Short('i').Bool()
	xDdForce         = xDropDatabase.Flag("force", "Suppress warning message.").Short('f').Bool()

	xChangeAccount = xApp.Command("change-password", "Change password for your service account.")
	xCaPassword    = xChangeAccount.Arg("password", "New password.").Required().String()

	xNewDatabase = xApp.Command("new-database", "Create a new SiriDB database.")
	xNdDatabase  = xNewDatabase.Flag(
		"db-name", "Database name. (Must be at least 2 and at most 20 characters. "+
			"Fist character must be a letter. "+
			"Last character must be a letter or number. "+
			"In between letters, numbers, hyphen and underscores are allowed)").Short('d').Required().String()
	xNdTimep   = xNewDatabase.Flag("time-precision", "Time precision for the new database.").Short('t').Default("ms").HintOptions("s", "ms", "us", "ns").String()
	xNdBufSize = xNewDatabase.Flag("buffer-size", "Buffer size for the new database.").Short('b').Default("1024").Int()
	xNdDuraNum = xNewDatabase.Flag("duration-num", "Number duration for the new database.").Short('N').Default("1w").String()
	xNdDuraLog = xNewDatabase.Flag("duration-log", "Log duration for the new database.").Short('L').Default("1d").String()

	xNewPool    = xApp.Command("new-pool", "Expand a SiriDB database with a new pool.")
	xNpDatabase = xNewPool.Flag("db-name", "Database name where you want to add the new pool to.").Short('d').Required().String()
	xNpUser     = xNewPool.Flag("db-user", "User with full privileges to the database.").Short('U').Required().String()
	xNpPassword = xNewPool.Flag("db-password", "Password for the database user.").Short('P').Required().String()
	xNpServer   = xNewPool.Flag("db-server", "SiriDB server address[:port]. Can be any server from the database you want to add a new pool to.").Short('S').Required().String()
	xNpForce    = xNewPool.Flag("force", "Suppress warning message.").Short('f').Bool()

	xNewReplica = xApp.Command("new-replica", "Expand a SiriDB pool with a new replica.")
	xNrDatabase = xNewReplica.Flag("db-name", "Database name where you want to add the new replica to.").Short('d').Required().String()
	xNrUser     = xNewReplica.Flag("db-user", "User with full privileges to the database.").Short('U').Required().String()
	xNrPassword = xNewReplica.Flag("db-password", "Password for the database user.").Short('P').Required().String()
	xNrServer   = xNewReplica.Flag("db-server", "SiriDB server address[:port]. Can be any server from the database you want to add a new replica to.").Short('S').Required().String()
	xNrPool     = xNewReplica.Flag("pool", "Pool number which you want to create the replica for.").Short('o').Required().Int()
	xNrForce    = xNewReplica.Flag("force", "Suppress warning message.").Short('f').Bool()
)

var globalSessions *session.Manager

const invalidServerAddress = "invalid server address: %s (valid examples: myserver.local, myserver.local:9000, ::1, [::1]:9000, etc...)\n"

func askForConfirmation(s string) bool {
	reader := bufio.NewReader(os.Stdin)

	for {
		fmt.Printf("%s [y/n]: ", s)

		response, err := reader.ReadString('\n')
		if err != nil {
			log.Fatal(err)
		}

		response = strings.ToLower(strings.TrimSpace(response))

		if response == "y" || response == "yes" {
			return true
		} else if response == "n" || response == "no" {
			return false
		}
	}
}

func getpass() string {
	var pass []byte
	var err error
	for len(pass) == 0 {
		fmt.Printf("Password: ")

		// Silent. For printing *'s use gopass.GetPasswdMasked()
		pass, err = gopass.GetPasswdMasked()
		if err != nil {
			println("error reading password")
			os.Exit(1)
		}
	}
	return string(pass)
}

func getHostAndPort(addr string) (string, uint16, error) {
	parts := strings.Split(addr, ":")
	// IPv4
	if len(parts) == 1 {
		return parts[0], 9000, nil
	}
	if len(parts) == 2 {
		u, err := strconv.ParseUint(parts[1], 10, 16)
		return parts[0], uint16(u), err
	}
	// IPv6
	if addr[0] != '[' {
		return fmt.Sprintf("[%s]", addr), 9000, nil
	}
	if addr[len(addr)-1] == ']' {
		return addr, 9000, nil
	}
	u, err := strconv.ParseUint(parts[len(parts)-1], 10, 16)
	addr = strings.Join(parts[:len(parts)-1], ":")
	return addr, uint16(u), err
}

func newDatabase(
	conn *siridb.Connection,
	account, password, dbname, timePrecision string,
	bufferSize int,
	durationNum, durationLog string) (interface{}, error) {
	var msg string
	options := make(map[string]interface{})

	options["dbname"] = dbname
	options["time_precision"] = timePrecision
	options["buffer_size"] = bufferSize
	options["duration_num"] = durationNum
	options["duration_log"] = durationLog

	_, err := conn.Manage(account, password, siridb.AdminNewDatabase, options)

	if err == nil {
		msg = fmt.Sprintf("successfully created database: %s\n", dbname)
	}
	return msg, err
}

func newPool(
	conn *siridb.Connection,
	account, password, dbname, serveraddr, username, userpw string,
	force bool) (interface{}, error) {
	var msg string
	server, port, err := getHostAndPort(serveraddr)
	if err != nil {
		return msg, fmt.Errorf(invalidServerAddress, serveraddr)
	}

	if !force {
		c := askForConfirmation("WARNING: It is not possible to undo this action!\nAre you sure you want to continue?")
		if !c {
			return msg, fmt.Errorf("cancelled by user")
		}
	}

	options := make(map[string]interface{})

	options["dbname"] = dbname
	options["host"] = server
	options["port"] = int(port)
	options["username"] = username
	options["password"] = userpw

	_, err = conn.Manage(account, password, siridb.AdminNewPool, options)

	if err == nil {
		msg = fmt.Sprintf("successfully created a pool for database: %s\n", dbname)
	}
	return msg, err
}

func newReplica(
	conn *siridb.Connection,
	account, password, dbname, serveraddr, username, userpw string,
	pool int, force bool) (interface{}, error) {
	var msg string
	server, port, err := getHostAndPort(serveraddr)
	if err != nil {
		return msg, fmt.Errorf(invalidServerAddress, serveraddr)
	}

	if !force {
		c := askForConfirmation("WARNING: It is not possible to undo this action!\nAre you sure you want to continue?")
		if !c {
			return msg, fmt.Errorf("cancelled by user")
		}
	}

	options := make(map[string]interface{})

	options["dbname"] = dbname
	options["host"] = server
	options["port"] = int(port)
	options["username"] = username
	options["password"] = userpw
	options["pool"] = pool

	_, err = conn.Manage(account, password, siridb.AdminNewReplica, options)

	if err == nil {
		msg = fmt.Sprintf("successfully created a replica for database: %s\n", dbname)
	}
	return msg, err
}

func getAccounts(conn *siridb.Connection, account, password string) (interface{}, error) {
	res, err := conn.Manage(account, password, siridb.AdminGetAccounts, nil)
	return res, err
}

func getDatabases(conn *siridb.Connection, account, password string) (interface{}, error) {
	res, err := conn.Manage(account, password, siridb.AdminGetDatabases, nil)
	return res, err
}

func getVersion(conn *siridb.Connection, account, password string) (interface{}, error) {
	res, err := conn.Manage(account, password, siridb.AdminGetVersion, nil)
	return res, err
}

func newAccount(conn *siridb.Connection, account, password, newAccnt, newPasswd string) (interface{}, error) {
	var msg string
	options := make(map[string]interface{})

	options["account"] = newAccnt
	options["password"] = newPasswd

	_, err := conn.Manage(account, password, siridb.AdminNewAccount, options)
	if err == nil {
		msg = fmt.Sprintf("successfully created service account: %s", newAccnt)
	}
	return msg, err
}

func dropAccount(conn *siridb.Connection, account, password, dropAccnt string) (interface{}, error) {
	var msg string
	options := make(map[string]interface{})

	options["account"] = dropAccnt

	_, err := conn.Manage(account, password, siridb.AdminDropAccount, options)
	if err == nil {
		msg = fmt.Sprintf("successfully dropped service account: %s", dropAccnt)
	}
	return msg, err
}

func dropDatabase(conn *siridb.Connection, account, password, dropDb string, ignoreOffline, force bool) (interface{}, error) {
	var msg string

	if !force {
		c := askForConfirmation("WARNING: This will drop the database on all servers in the cluster.\n\nIt is not possible to undo this action!\nAre you sure you want to continue?")
		if !c {
			return msg, fmt.Errorf("cancelled by user")
		}
	}

	options := make(map[string]interface{})

	options["database"] = dropDb
	options["ignore_offline"] = ignoreOffline

	_, err := conn.Manage(account, password, siridb.AdminDropDatabase, options)
	if err == nil {
		msg = fmt.Sprintf("successfully dropped database: %s", dropDb)
	}
	return msg, err
}

func changePassword(conn *siridb.Connection, account, password, changeAccount, newPassword string) (interface{}, error) {
	var msg string
	options := make(map[string]interface{})

	options["account"] = changeAccount
	options["password"] = newPassword

	_, err := conn.Manage(account, password, siridb.AdminChangePassword, options)
	if err == nil {
		msg = fmt.Sprintf("successfully changed password for service account: %s", changeAccount)
	}
	return msg, err
}

func logHandle(logCh chan string) {
	for {
		msg := <-logCh
		if *xVerbose {
			println(msg)
		}
	}
}

func initHTTP() error {
	var err error
	cf := new(session.ManagerConfig)
	cf.EnableSetCookie = true
	s := `{"cookieName":"siridbadminsessionid","gclifetime":3600}`

	if err = json.Unmarshal([]byte(s), cf); err != nil {
		return err
	}

	if globalSessions, err = session.NewManager("memory", cf); err != nil {
		return err
	}

	go globalSessions.GC()

	http.HandleFunc("/", handlerMain)
	http.HandleFunc("/js/bundle", handlerJsBundle)
	http.HandleFunc("/css/bootstrap", handlerBootstrapCSS)
	http.HandleFunc("/css/layout", handlerLayout)
	http.HandleFunc("/favicon.ico", handlerFaviconIco)
	http.HandleFunc("/img/siridb-large.png", handlerSiriDBLargePNG)
	http.HandleFunc("/img/siridb-small.png", handlerSiriDBSmallPNG)
	http.HandleFunc("/img/loader.gif", handlerLoaderGIF)
	http.HandleFunc("/css/font-awesome.min.css", handlerFontAwesomeMinCSS)
	http.HandleFunc("/fonts/FontAwesome.otf", handlerFontsFaOTF)
	http.HandleFunc("/fonts/fontawesome-webfont.eot", handlerFontsFaEOT)
	http.HandleFunc("/fonts/fontawesome-webfont.svg", handlerFontsFaSVG)
	http.HandleFunc("/fonts/fontawesome-webfont.ttf", handlerFontsFaTTF)
	http.HandleFunc("/fonts/fontawesome-webfont.woff", handlerFontsFaWOFF)
	http.HandleFunc("/fonts/fontawesome-webfont.woff2", handlerFontsFaWOFF2)
	http.HandleFunc("/version/fetch", handlerVersionFetch)
	http.HandleFunc("/accounts/fetch", handlerAccountsFetch)
	http.HandleFunc("/accounts/new", handlerAccountsNew)
	http.HandleFunc("/accounts/drop", handlerAccountsDrop)
	http.HandleFunc("/databases/fetch", handlerDatabasesFetch)
	http.HandleFunc("/databases/new-database", handlerDatabasesNewDatabase)
	http.HandleFunc("/databases/new-pool", handlerDatabasesNewPool)
	http.HandleFunc("/databases/new-replica", handlerDatabasesNewReplica)
	http.HandleFunc("/databases/drop", handlerDatabasesDrop)
	http.HandleFunc("/auth/fetch", handlerAuthFetch)
	http.HandleFunc("/auth/login", handlerAuthLogin)
	http.HandleFunc("/auth/logoff", handlerAuthLogoff)
	http.HandleFunc("/auth/change-password", handlerAuthChangePassword)
	return nil
}

var logChannel *chan string

func main() {
	var server, args string
	var err error
	var port uint16

	// parse arguments
	args, err = xApp.Parse(os.Args[1:])

	if *xVersion {
		fmt.Printf("Version: %s\n", AppVersion)
		os.Exit(0)
	}

	// suppress logging if not in verbose mode
	logCh := make(chan string)
	go logHandle(logCh)
	logChannel = &logCh

	if *xHTTP {
		if err != nil && strings.Compare(err.Error(), "required flag --user not provided") != 0 {
			fmt.Printf("%s\n", err)
			os.Exit(1)
		}

		initHTTP()

		fmt.Printf("Serving a graphical user interface on: http://0.0.0.0:%d\nPress CTRL+C to quit\n", *xHTTPPort)
		if err = http.ListenAndServe(fmt.Sprintf(":%d", *xHTTPPort), nil); err != nil {
			fmt.Printf("error: %s\n", err)
		}
	} else {
		if err != nil {
			fmt.Printf("%s\n", err)
			os.Exit(1)
		}

		server, port, err = getHostAndPort(*xServer)
		if err != nil {
			fmt.Printf(invalidServerAddress, *xServer)
			os.Exit(1)
		}

		// ask password if not supplied
		if len(*xPassword) == 0 {
			pass := getpass()
			xPassword = &pass
		}

		conn := siridb.NewConnection(server, port)
		conn.LogCh = *logChannel

		var msg interface{}

		switch kingpin.MustParse(args, err) {
		case xGetVersion.FullCommand():
			msg, err = getVersion(conn, *xAccount, *xPassword)
		case xGetAccounts.FullCommand():
			msg, err = getAccounts(conn, *xAccount, *xPassword)
		case xGetDatabases.FullCommand():
			msg, err = getDatabases(conn, *xAccount, *xPassword)
		case xNewAccount.FullCommand():
			msg, err = newAccount(conn, *xAccount, *xPassword, *xNaAccount, *xNaPassword)
		case xDropAccount.FullCommand():
			msg, err = dropAccount(conn, *xAccount, *xPassword, *xDaAccount)
		case xDropDatabase.FullCommand():
			msg, err = dropDatabase(conn, *xAccount, *xPassword, *xDdDatabase, *xDdIgnoreOffline, *xDdForce)
		case xChangeAccount.FullCommand():
			msg, err = changePassword(conn, *xAccount, *xPassword, *xAccount, *xCaPassword)
		case xNewDatabase.FullCommand():
			msg, err = newDatabase(conn, *xAccount, *xPassword, *xNdDatabase, *xNdTimep, *xNdBufSize, *xNdDuraNum, *xNdDuraLog)
		case xNewPool.FullCommand():
			msg, err = newPool(conn, *xAccount, *xPassword, *xNpDatabase, *xNpServer, *xNpUser, *xNpPassword, *xNpForce)
		case xNewReplica.FullCommand():
			msg, err = newReplica(conn, *xAccount, *xPassword, *xNrDatabase, *xNrServer, *xNrUser, *xNrPassword, *xNrPool, *xNrForce)
		}

		conn.Close()

		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		fmt.Println(msg)
	}
}
